source/adapters/hip/memory.cpp

```diff
@@ -282,16 +282,34 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemGetInfo(ur_mem_handle_t hMemory,
 /// \param[out] phNativeMem Set to the native handle of the UR mem object.
 ///
 /// \return UR_RESULT_SUCCESS
-UR_APIEXPORT ur_result_t UR_APICALL urMemGetNativeHandle(ur_mem_handle_t,
-                                                         ur_native_handle_t *) {
-  // FIXME: there is no good way of doing this with a multi device context.
-  // If we return a single pointer, how would we know which device's allocation
-  // it should be?
-  // If we return a vector of pointers, this is OK for read only access but if
-  // we write to a buffer, how would we know which one had been written to?
-  // Should unused allocations be updated afterwards? We have no way of knowing
-  // any of these things in the current API design.
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+UR_APIEXPORT ur_result_t UR_APICALL
+urMemGetNativeHandle(ur_mem_handle_t hMem, ur_device_handle_t Device,
+                     ur_native_handle_t *phNativeMem) {
+  if (!Device)
+    return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+#if defined(__HIP_PLATFORM_NVIDIA__)
+  if (sizeof(BufferMem::native_type) > sizeof(ur_native_handle_t)) {
+    // Check that all the upper bits that cannot be represented by
+    // ur_native_handle_t are empty.
+    // NOTE: The following shift might trigger a warning, but the check in the
```

Comment:
If you want to do bitwise arithmetic on a pointer, and not have the compiler - or language lawyers - complain, you need to convert it first to a suitable type, such as `uintptr_t`.

Reply:
Also, can you document why we need to check these bits? Is this something to do with how HIP encodes device info? If so: is it documented by NVIDIA?

Reply:
This was pre-existing code from before the multi device context PR so I really don't know why we needed to do this.

Reply:
Aha https://github.com/intel/llvm/pull/6207

Reply:
Oh right. It's a little suspect, so we can track down the meaning and update in another patch that'd be good.